tests/unit/__init__.py

```diff
@@ -0,0 +1,172 @@
+import logging
+import operator
+import re
+from typing import Any, Callable, Iterable, List, Tuple
+
+import _pytest.logging
+
+
+class PatternMatching:
+    def __init__(self, pattern: str, method: str) -> None:
+        self.pattern = pattern
+        self._compiled_pattern = re.compile(pattern)
+        self.method = getattr(self._compiled_pattern, method)
+
+    def __repr__(self) -> str:
+        return f'<{self.__class__.__name__}: "{self.pattern}">'
+
+
+class MATCH(PatternMatching):
+    """
+    Wrap a string with this class, to use it as a regular expression when matching log records.
+
+    :py:class:`SEARCH` applies to any place within the string, while ``MATCH`` must match from the
+    beginning of the string.
+
+    .. code-block:: python
+
+       assert_log(message=MATCH('an exception .+ was raised'))
+    """
+
+    def __init__(self, pattern: str) -> None:
+        super().__init__(pattern, 'match')
+
+
+class SEARCH(PatternMatching):
+    """
+    Wrap a string with this class, to use it as a regular expression when searching for log
+    records.
+
+    ``SEARCH`` applies to any place within the string, while :py:class:`MATCH` must match from the
+    beginning of the string.
+
+    .. code-block:: python
+
+       assert_log(message=SEARCH('an exception .+ was raised'))
+    """
+
+    def __init__(self, pattern: str) -> None:
+        super().__init__(pattern, 'search')
+
+
+def _assert_log(
+        caplog: _pytest.logging.LogCaptureFixture,
+        evaluator: Callable[[Iterable[Any]], bool] = any,
+        not_: bool = False,
+        **tests: Any
+        ) -> None:
+    """
+    Assert log contains a record - logged message - with given properties. Those are specified as
+    keyword parameters: :py:class:`logging.LogRecords` properties are allowed names, parameter
+    values are the expected values.
+
+    .. code-block:: python
+
+       assert_log(message='everything went well', levelno=logging.INFO)
+       assert_log(message='things broke down', levelno=logging.ERROR)
+       assert_log(message=MATCH('user .+ logged in'), levelno=logging.INFO)
+
+    :param caplog: Pytest's `caplog` fixture.
+    :param evaluator: a callable reducing a given list of booleans into a single boolean. It is
+        used to evaluate whether the search for matching record was successfull: each record is
+        tested, and results of these per-record tests are passed to `evaluator` for the final
+        decision.
```

Comment:
Please, add a description of the `not_` parameter. Or perhaps we could rename it to `present` or `should_be_present`, `expected` or something like this?

Reply:
Changed in 6d14224c36569188909070dbdd7f1a5fe7fcbf6f